src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.cs

```diff
@@ -121,142 +116,162 @@ async Task<(ImmutableArray<SerializableImportCompletionItem>, StatisticCounter)>
 
             counter.GetSymbolTicks = Environment.TickCount - ticks;
 
             return (items, counter);
         }
 
-        private static async Task<GetIndicesResult> TryGetIndicesAsync(
-            Project currentProject,
-            bool forceIndexCreation,
+        private static async Task<ImmutableArray<SerializableImportCompletionItem>> GetExtensionMethodItemsAsync(
+            Document document,
+            ITypeSymbol receiverTypeSymbol,
+            ImmutableArray<string> targetTypeNames,
+            GetIndicesResult indices,
+            int position,
+            ISet<string> namespaceFilter,
+            StatisticCounter counter,
             CancellationToken cancellationToken)
         {
-            var solution = currentProject.Solution;
-            var cacheService = GetCacheService(solution.Workspace);
-            var graph = currentProject.Solution.GetProjectDependencyGraph();
-            var relevantProjectIds = graph.GetProjectsThatThisProjectTransitivelyDependsOn(currentProject.Id)
-                                        .Concat(currentProject.Id);
+            if (!indices.HasResult)
+            {
+                return ImmutableArray<SerializableImportCompletionItem>.Empty;
+            }
 
-            using var syntaxDisposer = ArrayBuilder<CacheEntry>.GetInstance(out var syntaxBuilder);
-            using var symbolDisposer = ArrayBuilder<SymbolTreeInfo>.GetInstance(out var symbolBuilder);
+            var currentProject = document.Project;
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var currentCompilation = semanticModel.Compilation;
+            var currentAssembly = currentCompilation.Assembly;
 
-            foreach (var projectId in relevantProjectIds)
-            {
-                var project = solution.GetProject(projectId);
-                if (project == null || !project.SupportsCompilation)
-                {
-                    continue;
-                }
+            using var _1 = ArrayBuilder<SerializableImportCompletionItem>.GetInstance(out var builder);
+            using var _2 = PooledDictionary<INamespaceSymbol, string>.GetInstance(out var namespaceNameCache);
 
-                // By default, don't trigger index creation except for documents in current project.
-                var loadOnly = !forceIndexCreation && projectId != currentProject.Id;
-                var cacheEntry = await GetCacheEntryAsync(project, loadOnly, cacheService, cancellationToken).ConfigureAwait(false);
+            // Get extension method items from source
+            foreach (var (project, syntaxIndex) in indices.SyntaxIndices!)
+            {
+                var filter = CreateAggregatedFilter(targetTypeNames, syntaxIndex);
+                var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+                var assembly = compilation.Assembly;
```

Comment:
so this can be expensive.  i forget how this feature works.  is it gated on time?

Reply:
No, it will run to the end as long as we have the index available.

Reply:
Also I'm thinking maybe more caching can be done to improve perf, e.g. cache SymbolIds once we've got the symbols from assemblies

Reply:
@CyrusNajmabadi  Do you know how compilation cache works (if there's one)? In what case a built final compilation is kicked off the cache?

Reply:
Tagging @jasonmalinowski

Reply:
in the future we should optimize this to search the current project first.